apps/src/templates/progress/progressHelpers.js

```diff
@@ -180,6 +241,176 @@ export const processedLevel = level => {
     levelNumber: level.kind === LevelKind.unplugged ? undefined : level.title,
     isConceptLevel: level.is_concept_level,
     bonus: level.bonus,
-    sublevels: level.sublevels
+    sublevels:
+      level.sublevels &&
+      level.sublevels.map(level => processedLevel(level, true))
+  };
+};
+
+/**
+ * Given a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress,
+ * extracts the result, appropriately discerning a locked/submitted
+ * result for certain levels.
+ */
+const getLevelResult = serverProgress => {
+  if (serverProgress.status === LevelStatus.locked) {
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || TestResults.NO_TESTS_RUN;
+};
+
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
+    paired: serverObject.paired || false,
+    timeSpent: serverObject.time_spent || 0
   };
 };
+
+export const processServerStudentProgress = serverStudentProgress => {
```

Comment:
Many of these seem to have similar names and do similar things. Would you mind adding some comments to help differentiate when we might want to choose one or another? Perhaps a visual of how the data is transformed would be helpful.

Reply:
yes many of these started out as private helper functions, but since we process data at different levels of granularity in various parts of our code they all ended up getting exported. i've added comments, but imo which one to use is pretty apparent in context depending on the format of the data we're working with. lmk if this still needs more comments after reviewing part 2.

Reply:
👍